core/nativeMain/src/kotlinx/io/buffer/NativeBuffer.kt

```diff
@@ -3,22 +3,35 @@
 package kotlinx.io.buffer
 
 import kotlinx.cinterop.*
-import kotlinx.io.bits.internal.utils.*
-import kotlin.native.concurrent.SharedImmutable
+import kotlin.native.concurrent.*
 
-public actual class Buffer constructor(val pointer: CPointer<ByteVar>, actual inline val size: Int) {
+public actual class Buffer constructor(
+    val array: ByteArray,
+    inline val offset: Int = 0,
+    actual inline val size: Int = array.size - offset
+) {
     init {
         requirePositiveIndex(size, "size")
     }
 
-    public actual inline fun loadByteAt(index: Int): Byte = pointer[assertIndex(index, 1)]
+    public actual inline fun loadByteAt(index: Int): Byte = array[assertIndex(offset + index, 1)]
 
     public actual inline fun storeByteAt(index: Int, value: Byte) {
-        pointer[assertIndex(index, 1)] = value
+        array[assertIndex(offset + index, 1)] = value
     }
 
     public actual companion object {
         @SharedImmutable
-        public actual val EMPTY: Buffer = Buffer(nativeHeap.allocArray(0), 0)
+        public actual val EMPTY: Buffer = Buffer(ByteArray(0))
     }
 }
+
+inline fun <R> Buffer.usePointer(block: (pointer: CPointer<ByteVar>) -> R) = array.usePinned {
```

Comment:
1) Should it really be public?
2) Why all our primitives (`load*`) are implemented on top of unsafe pinned usage instead of properly and safely working with `ByteArray`?

Reply:
1. It's the only way to implement adapters for syscalls
2. I don't see any issues safeness here and the implementation is correct, so could we consider keeping it as it is until the benchmarking phase?

Reply:
>It's the only way to implement adapters for syscalls

Good. Then please make it explicitly public, document and test